collector/adfs.go

```diff
@@ -185,5 +442,197 @@ func (c *adfsCollector) Collect(ctx *ScrapeContext, ch chan<- prometheus.Metric)
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.oAuthClientWindowsIntegratedAuthenticationFailures,
+		prometheus.CounterValue,
+		adfsData[0].OAuthClientWindowsAuthnFailures,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.oAuthClientWindowsIntegratedAuthentications,
+		prometheus.CounterValue,
+		adfsData[0].OAuthClientWindowsAuthentications,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.oAuthLogonCertificateRequestFailures,
+		prometheus.CounterValue,
+		adfsData[0].OAuthLogonCertRequestFailures,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.oAuthLogonCertificateTokenRequests,
+		prometheus.CounterValue,
+		adfsData[0].OAuthLogonCertTokenRequests,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.oAuthPasswordGrantRequestFailures,
+		prometheus.CounterValue,
+		adfsData[0].OAuthPasswordGrantRequestFailures,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.oAuthPasswordGrantRequests,
+		prometheus.CounterValue,
+		adfsData[0].OAuthPasswordGrantRequests,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.oAuthTokenRequests,
+		prometheus.CounterValue,
+		adfsData[0].OAuthTokenRequests,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.samlPTokenRequests,
+		prometheus.CounterValue,
+		adfsData[0].SAMLPTokenRequests,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.ssoAuthenticationFailures,
+		prometheus.CounterValue,
+		adfsData[0].SSOAuthenticationFailures,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.ssoAuthentications,
+		prometheus.CounterValue,
+		adfsData[0].SSOAuthentications,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.wsfedTokenRequests,
+		prometheus.CounterValue,
+		adfsData[0].WSFedTokenRequests,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.wstrustTokenRequests,
+		prometheus.CounterValue,
+		adfsData[0].WSTrustTokenRequests,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.upAuthenticationFailures,
+		prometheus.CounterValue,
+		adfsData[0].UsernamePasswordAuthnFailures,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.upAuthentications,
+		prometheus.CounterValue,
+		adfsData[0].UsernamePasswordAuthentications,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.externalAuthenticationFailures,
+		prometheus.CounterValue,
+		adfsData[0].ExternalAuthNFailures,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.externalAuthentications,
+		prometheus.CounterValue,
+		adfsData[0].ExternalAuthentications,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.artifactDBFailures,
+		prometheus.CounterValue,
+		adfsData[0].ArtifactDBFailures,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.avgArtifactDBQueryTime,
+		prometheus.GaugeValue,
+		adfsData[0].AvgArtifactDBQueryTime*math.Pow(10, -8),
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.configDBFailures,
+		prometheus.CounterValue,
+		adfsData[0].ConfigDBFailures,
+	)
+
+	ch <- prometheus.MustNewConstMetric(
+		c.avgConfigDBQueryTime,
+		prometheus.GaugeValue,
+		adfsData[0].AvgConfigDBQueryTime*math.Pow(10, -8),
+	)
```

Comment:
Metric name and description don't match the documentation. Documentation states this metric is the duration of the most recent query, while the metric in the code indicates an average of all queries.

Reply:
I believe this is another one of those things where Perflib is doing something behind the scenes to the value. If you graph out that value over time (top graph below), it is ever increasing and only resets during a reboot, which would make me guess that perflib is summing the value, rather than showing an average. My intuition tells me that it's not taking 5 seconds to read from the database, and 5-10ms seems to be more realistic (when `increase` is applied).

Perhaps we should make it `_total` so that way folks know to wrap it in an `increase()` (below graph).

![image](https://user-images.githubusercontent.com/9043657/148830822-2cc3d294-f5c9-4bde-9fcc-c38b4e588f65.png)

Reply:
Yes, exposing the metrics as a counter with the `_total` suffix would be the correct option here.
Perflib can be "interesting" in that you can't be entirely certain if a metric is a gauge or counter until directly observing it.

Reply:
Updated metrics and description to `db_config_query_time_seconds_total` and `Accumulator of time taken for a configuration database query` respectively.